cpp/src/community/detail/mis_impl.cuh

```diff
@@ -117,22 +104,47 @@ rmm::device_uvector<vertex_t> compute_mis(
-    // and for MG extension, select from disributed array remaining vertices
-    thrust::shuffle(
-      handle.get_thrust_policy(), remaining_vertices.begin(), remaining_vertices.end(), g);
 
-    vertex_t nr_candidates =
-      (remaining_vertices.size() < 1024)
-        ? remaining_vertices.size()
-        : std::min(static_cast<vertex_t>((0.50 + 0.25 * loop_counter) * remaining_vertices.size()),
-                   static_cast<vertex_t>(remaining_vertices.size()));
+    vertex_t nr_remaining_vertices_to_check = remaining_vertices.size();
+    if (multi_gpu) {
+      nr_remaining_vertices_to_check = host_scalar_allreduce(handle.get_comms(),
+                                                             nr_remaining_vertices_to_check,
+                                                             raft::comms::op_t::SUM,
+                                                             handle.get_stream());
+    }
+
+    vertex_t nr_candidates = (nr_remaining_vertices_to_check < 1024)
+                               ? nr_remaining_vertices_to_check
+                               : std::min(static_cast<vertex_t>((0.50 + 0.25 * loop_counter) *
+                                                                nr_remaining_vertices_to_check),
+                                          nr_remaining_vertices_to_check);
+
+    // FIXME: (nr_remaining_vertices_to_check < 1024)? avoid calling select_random_vertices
+    auto d_sampled_vertices =
+      cugraph::select_random_vertices(handle,
+                                      graph_view,
+                                      std::make_optional(raft::device_span<vertex_t const>{
+                                        remaining_vertices.data(), remaining_vertices.size()}),
+                                      rng_state,
+                                      nr_candidates,
+                                      false,
+                                      true);
+
+    rmm::device_uvector<vertex_t> non_candidate_vertices(
+      remaining_vertices.size() - d_sampled_vertices.size(), handle.get_stream());
+
+    thrust::set_difference(handle.get_thrust_policy(),
+                           remaining_vertices.begin(),
+                           remaining_vertices.end(),
+                           d_sampled_vertices.begin(),
+                           d_sampled_vertices.end(),
+                           non_candidate_vertices.begin());
 
     // Set temporary ranks of non-candidate vertices to std::numeric_limits<vertex_t>::lowest()
     thrust::for_each(
       handle.get_thrust_policy(),
-      remaining_vertices.begin(),
-      remaining_vertices.end() - nr_candidates,
+      non_candidate_vertices.begin(),
+      non_candidate_vertices.end(),
       [temporary_ranks =
          raft::device_span<vertex_t>(temporary_ranks.data(), temporary_ranks.size()),
        v_first = graph_view.local_vertex_partition_range_first()] __device__(auto v) {
```

Comment:
Regarding the comment below

```

    //
    // Find maximum rank outgoing neighbor for each vertex
    // (In case of Leiden decision graph, each vertex has at most one outgoing edge)
    //
```

If this code is for general MIS and you plan to expose this in the public API, better remove any Leiden specific comments.